Sources/ObliviousHTTP/BHTTPSerializer.swift

```diff
@@ -88,18 +160,82 @@ public struct BHTTPSerializer {
             buffer.writeVarintPrefixedString(name)
             buffer.writeVarintPrefixedString(value)
         }
-        // This is technically a varint but we can skip the check there because we know it can always encode in one byte.
-        buffer.writeInteger(UInt8(0))
+        buffer.writeInteger(UInt8(0))  // End of field section
+    }
+
+    private mutating func serializeKnownLengthFieldSection(into buffer: inout ByteBuffer) {
+        buffer.writeVarintPrefixedImmutableBuffer(self.fieldSectionBuffer)
+        self.fieldSectionBuffer.clear()
     }
 
+    private mutating func stackKnownLengthFieldSection(_ fields: HTTPHeaders) {
+        for (name, value) in fields {
+            self.fieldSectionBuffer.writeVarintPrefixedString(name)
+            self.fieldSectionBuffer.writeVarintPrefixedString(value)
+        }
+    }
 }
 
+// Enum definitions for message, states, and types.
 extension BHTTPSerializer {
-    /// Types of message for binary http serilaisation
+    // Finite State Machine for managing transitions in BHTTPSerializer.
+    private class BHTTPSerializerFSM {
+        var currentState: BHTTPSerializerState
+
+        init(initialState: BHTTPSerializerState) {
+            self.currentState = initialState
+        }
+
+        // Transition to a new state, respecting the state machine constraints.
+        func transition(to state: BHTTPSerializerState) throws {
+            guard let allowedTransitions = Self.validTransitions[currentState] else {
+                throw ObliviousHTTPError.unexpectedHTTPMessageSection(state: currentState.rawValue)
+            }
+
+            guard allowedTransitions.contains(state) else {
+                throw ObliviousHTTPError.unexpectedHTTPMessageSection(state: currentState.rawValue)
+            }
+
+            currentState = state
+        }
+
+        // Define a dictionary to map current states to allowed transitions
+        private static let validTransitions: [BHTTPSerializerState: Set<BHTTPSerializerState>] = [
```

Comment:
I might suggest replacing this dictionary with a `switch` statement. Hashing the enumeration is going to be a lot more expensive than switching over it.